Write intercepted mapping to generated/metadata during compilation

pmclain opened the pull request:
### Description (*)
This prevents the intercepted cache from being cleared following compilation
by writing the compiled interception cache to `generated/metadata`. The
mapping is stored in the frontend cache when the mapping metadata file is
not present, ie developer, default modes.

### Fixed Issues (if relevant)
1. magento/magento2#17680 : [2.2.3+][CE/EE][Performance] Interception Cache compilation useless

### Manual testing scenarios (*)
1. Install clean Magento 2.3-deveop Community/Commerce with sample data
2. Add profiler start/stop for next methods:
    * \Magento\Framework\Interception\Config\Config::__construct
    * \Magento\Framework\Interception\PluginList\PluginList::_loadScopedData
    * \Magento\Framework\ObjectManager\Config\Reader\Dom::read
    Note: Profiler start/stop can be added like this:
    ```
     \Magento\Framework\Profiler::start(__METHOD__);
     //Method code here
     \Magento\Framework\Profiler::stop(__METHOD__);
    ```
3. Enable CSV profiler - execute "bin/magento dev:profiler:enable csvfile"
4. Enable PROD mode - execute "bin/magento deploy:mode:set production". This will run setup:di:compile + static:content:deploy
Note step # 4 of di:compile - "Interception cache generation".
5. Run cache:flush
6. Try to open any URL of your store.
7. Inspect profiler results in "var/log/profiler.csv".

### Expected result
1. Our custom profiler records should be executed in less than second together:
    * `Config::__construct`
    * `PluginList::_loadScopedData`
    * `Dom::read`
2. Interception config should be compiled and stored to filesystem like other compiled configs in `generated/metadata` folder. E.g. by using next calls:
    * `\Magento\Setup\Module\Di\Compiler\Config\Writer\Filesystem::write`
    * `\Magento\Framework\App\ObjectManager\ConfigLoader\Compiled::load`

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

pmclain committed:
Write intercepted mapping to generated/metadata during compilation

This prevents the intercepted cache from being cleared following compilation.
The mapping is stored in the frontend cache when the mapping metadata file is
not present, ie developer, default modes.

Fixes #17680

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=18648) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=18648) before we can accept your contribution.<br/>

magento-engcom-team commented:
Hi @pmclain. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

pmclain committed:
Move ObjectManager Config Writer to Framework

The ObjectManager Config Writer class currently lives in the Setup module and
is needed in the Framework when writing the interception config to the
filesystem during compilation. This commit adds the classes to the Framework
and replaces any existing usage with the new interface/class.

slavvka commented on dev/tests/static/testsuite/Magento/Test/Legacy/_files/obsolete_classes.php:
```diff
@@ -3350,6 +3350,8 @@
     ['Magento\Tools\Di\Compiler\Config\Reader', 'Magento\Setup\Module\Di\Compiler\Config\Reader'],
     ['Magento\Tools\Di\Compiler\Config\Writer\Filesystem', 'Magento\Setup\Module\Di\Compiler\Config\Writer\Filesystem'],
     ['Magento\Tools\Di\Compiler\Config\WriterInterface', 'Magento\Setup\Module\Di\Compiler\Config\WriterInterface'],
+    ['Magento\Setup\Module\Di\Compiler\Config\WriterInterface', 'Magento\Framework\App\ObjectManager\ConfigWriterInterface'],
```
You shouldn't have added classes to obsoletes unless they are physically removed

slavvka commented on dev/tests/static/testsuite/Magento/Test/Legacy/_files/obsolete_classes.php:
```diff
@@ -3350,6 +3350,8 @@
     ['Magento\Tools\Di\Compiler\Config\Writer\Filesystem', 'Magento\Setup\Module\Di\Compiler\Config\Writer\Filesystem'],
     ['Magento\Tools\Di\Compiler\Config\WriterInterface', 'Magento\Setup\Module\Di\Compiler\Config\WriterInterface'],
+    ['Magento\Setup\Module\Di\Compiler\Config\WriterInterface', 'Magento\Framework\App\ObjectManager\ConfigWriterInterface'],
+    ['Magento\Setup\Module\Di\Compiler\Config\Writer\Filesystem', 'Magento\Framework\App\ObjectManager\ConfigWriter\Filesystem'],
```
The same

slavvka commented on lib/internal/Magento/Framework/App/ObjectManager/ConfigWriter/Filesystem.php:
```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ *
```
Please remove empty line

slavvka commented on lib/internal/Magento/Framework/App/ObjectManager/ConfigWriter/Filesystem.php:
```diff
@@ -0,0 +1,59 @@
+use Magento\Framework\App\Filesystem\DirectoryList;
+use Magento\Framework\App\ObjectManager\ConfigWriterInterface;
+
+class Filesystem implements ConfigWriterInterface
```
Please add a description to the class or at least inherit it from the interface with `@inheritdoc`

slavvka replied:
I meant this one

slavvka commented on lib/internal/Magento/Framework/App/ObjectManager/ConfigWriterInterface.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+/**
+ *
```
Please remove empty line

slavvka commented on lib/internal/Magento/Framework/App/ObjectManager/ConfigWriterInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+namespace Magento\Framework\App\ObjectManager;
+
+interface ConfigWriterInterface
```
Please add a description to the interface

slavvka replied:
I meant this one

slavvka commented on lib/internal/Magento/Framework/Interception/Config/Config.php:
```diff
@@ -188,4 +191,64 @@ public function hasPlugins($type)
+    }
+
+    /**
+     * @return array|false
```
Please add a description

slavvka replied:
I meant this one

slavvka commented on lib/internal/Magento/Framework/Interception/Config/Config.php:
```diff
@@ -188,4 +191,64 @@ public function hasPlugins($type)
+    }
+
+    /**
+     * @return bool
```
Please add a description

slavvka replied:
I meant this one

slavvka commented on setup/src/Magento/Setup/Module/Di/Compiler/Config/Writer/Filesystem.php:
```diff
@@ -10,6 +10,10 @@
 use Magento\Setup\Module\Di\Compiler\Config\WriterInterface;
 
+/**
+ * @deprecated
```
Please mention the reason of deprecating the class

slavvka commented on setup/src/Magento/Setup/Module/Di/Compiler/Config/WriterInterface.php:
```diff
@@ -9,7 +9,8 @@
 /**
  * Interface \Magento\Setup\Module\Di\Compiler\Config\WriterInterface
- *
+ * @deprecated
```
Please mention the reason of deprecating the class

slavvka reviewed (changes requested):
Hey @pmclain thank you for this valuable contribution! Could you please make requested changes?

slavvka commented:
@pmclain You change very important part of Magento. Please consider covering your changes with appropriate tests (preferably integration).

pmclain commented:
@slavvka Thanks for the feedback. I've made the requested changes. I was able to add an integration test for the config writer class moved into the framework from the setup module. 
I'm at a bit of a loss on how to further extend the existing tests for the config class. The current interception tests instantiate and depend on the modified config class https://github.com/magento/magento2/blob/b0e60f77e9a4f0ec0a03c0b561fb419767382f27/dev/tests/integration/testsuite/Magento/Framework/Interception/AbstractPlugin.php#L44 Do you have any possible guidance on what tests should be added here?

slavvka commented:
@pmclain Thank you for the changes. 
As far as I can see `dev/tests/integration/testsuite/Magento/Framework/Interception/AbstractPlugin.php` is created as base class for testing Interception framework using plugins. I think that in your case it would be good to create a different integration test that test the process of loading/caching the configuration using the changed class `\Magento\Framework\Interception\Config\Config`. Also it would be good to extend the unit test `\Magento\Framework\Interception\Test\Unit\Config\ConfigTest` to cover scenarios of loading/caching. Does it make sense?

pmclain commented:
@slavvka That's very clear. Thank you for the guidance!

pmclain committed:
Add Interception Config Integration Test

Test load logic added for persisting compiled cache data in
`generated/metadata`

slavvka commented:
@pmclain Thank you for the test. It looks good. Could you please fix my comments I did before?

slavvka commented:
@pmclain no, I managed to see them. Today GitHub is experiencing issues.

magento-engcom-team commented:
Hi @slavvka, thank you for the review.
[ENGCOM-3243](https://jira.corp.magento.com/browse/ENGCOM-3243) has been created to process this Pull Request

slavvka commented:
Hey @pmclain. Another issue arose in static tests:
```Magento\Test\Php\LiveCodeTest::testCodeMess
PHP Code Mess has found error(s):
/var/www/html/lib/internal/Magento/Framework/Interception/Config/Config.php:13	The class Config has a coupling between objects value of 13. Consider to reduce the number of dependencies under 13.
/var/www/html/lib/internal/Magento/Framework/Interception/Config/Config.php:105	The method __construct has 10 parameters. Consider reducing the number of parameters to less than 10.
```
Could you please consider refactoring the solution to maintain the code complexity low?

slavvka commented:
And please change class usage to interface in the test:
```dev/tests/integration/testsuite/Magento/Framework/Interception/Config/ConfigTest.php
Class 'Magento\Framework\Serialize\Serializer\Serialize' is restricted in /var/www/html/dev/tests/integration/testsuite/Magento/Framework/Interception/Config/ConfigTest.php. Suggested replacement: Magento\Framework\Serialize\SerializerInterface```

pmclain committed:
Update serializer used in integration test

Replace serialize with it's interface. Also passes the serializer when
instantiating the test subject. This ensures the correct implementation is used
becuase the interface preference and fallback in the constructor do not match.

pmclain commented:
@slavvka Updated the integration test. I'm looking at the constructor, as it should be possible to refactor out some dependencies.

pmclain committed:
Separate interception config logic

Moves the logic responsible for loading, saving, clearing the interception
config cache into it's own class

pmclain commented:
@slavvka I separated the config load/save logic into it's own class which removed a number of dependancies from the `Config` class. This removed the dependency on the cache and serializer classes so I made their constructor arguments optional and removed the type-hints. To be honest I hadn't removed dependencies before, but It looked like that was what outlined in the [contribution guides](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/#refactoring-classes-that-reach-limit-of-coupling-between-objects).

slavvka commented on lib/internal/Magento/Framework/Interception/Config/Config.php:
```diff
@@ -74,46 +69,46 @@ class Config implements \Magento\Framework\Interception\ConfigInterface
         \Magento\Framework\Config\ReaderInterface $reader,
         \Magento\Framework\Config\ScopeListInterface $scopeList,
-        \Magento\Framework\Cache\FrontendInterface $cache,
+        $cache = null,
```
No, please dont change the signature of the constructor even if it is not used anymore

slavvka replied:
Generally you cannot add optional argument between required ones.

slavvka replied:
And I am sure that arguments should be marked as deprecated in order to remove the later. I cannot help you just now but you could search on Devdocs. Please ask me if you can't find that

slavvka commented on app/etc/di.xml:
```diff
@@ -407,11 +407,16 @@
+    <type name="Magento\Framework\Interception\Config\CacheManager">
+        <arguments>
+            <argument name="cache" xsi:type="object">Magento\Framework\App\Cache\Type\Config</argument>
+            <argument name="configWriter" xsi:type="object">Magento\Framework\App\ObjectManager\ConfigWriter\Filesystem</argument>
```
Why did you introduce an explicit type here instead of adding a preference for `\Magento\Framework\App\ObjectManager\ConfigWriterInterface`?

slavvka commented on app/etc/di.xml:
```diff
@@ -407,11 +407,16 @@
     </type>
     <type name="Magento\Framework\Interception\Config\Config">
         <arguments>
-            <argument name="cache" xsi:type="object">Magento\Framework\App\Cache\Type\Config</argument>
```
I don't think that we need to remove it. Try to find how to mark them as deprecated in Devdocs.

slavvka commented:
And please pay your attention at failed integration tests

pmclain commented:
@slavvka I'm still working on figuring out why the two MessageQueue deprecated use case tests fail after separating the cache load logic. I've had no luck reproducing the failure locally (limited the tests to `Framework/Interception` and `Framework/MessageQueue`. It only occurs when running the full batch from index 2. What are your thoughts on dropping this commit https://github.com/magento/magento2/pull/18648/commits/576135f77d2ac2695e7b3277c0e80991a4896097? The goal of that commit was to reduce the number of dependencies, but I've honestly hit a bit of a wall. 

> Hey @pmclain. Another issue arose in static tests:
> 
> ```
> PHP Code Mess has found error(s):
> /var/www/html/lib/internal/Magento/Framework/Interception/Config/Config.php:13	The class Config has a coupling between objects value of 13. Consider to reduce the number of dependencies under 13.
> /var/www/html/lib/internal/Magento/Framework/Interception/Config/Config.php:105	The method __construct has 10 parameters. Consider reducing the number of parameters to less than 10.
> ```
> Could you please consider refactoring the solution to maintain the code complexity low?

slavvka commented:
@pmclain we cannot accept increased complexity since it breaks our technical guideline. Please try to fix the test.

pmclain commented:
@slavvka Found it. https://github.com/magento/magento2/pull/18648/files#diff-6290c7957e4aa8c3066e04f8115ad536 instantiates the interception config class for testing plugin behaviour and was using a mock for the `CacheInterface`. The cache mock let the test force a reload of the interception config during setup, which was cold on the first setup in test batch 2, and prevent the test config from writing to the actual cache.
After https://github.com/magento/magento2/commit/576135f77d2ac2695e7b3277c0e80991a4896097 the interaction with the cache was moved to the `CacheManager` class and a live instance of that class was being instantiated during the tests since the new dependancy was backwards compatible. The all the interception tests passed due to the cold cache in the first test, but the tests' interception config was saved to the cache.
Seems a bit like a flaw in the test framework that there is not a proper way to ensure isolation for the cache backend similar to what is possible for the database.

pmclain commented:
The integration tests are failing on the target branch. Looks like the test distribution https://github.com/magento/magento2/blob/2.3-develop/dev/travis/before_script.sh#L16-L18 needs be adjusted to avoid timing out since https://github.com/magento/magento2/commit/da1a96c32eb105606453e89a5ff70b550b4f20f7

pmclain commented:
@slavvka Should I rebalance the integration test mix so they can complete in travis without timing out? I've been waiting for the mix to change in 2.3-develop so I can rebase all my pending PRs without introducing a merge conflict by including an edit to the travis script.

slavvka commented:
@pmclain yes, you could try that, thank you!

pmclain committed:
rebalance integration tests to prevent travis timeout

pmclain commented:
@slavvka Adjusted the test allocation here https://github.com/magento/magento2/pull/18648/commits/4c495f2faa29edcaeebfe1ca1691754f3ef41b55

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed docblock

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed static test

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed docblock

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

Fixed dockblock

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed docblock

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed docblock

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed docblock

slavvka committed:
ENGCOM-3243: Write intercepted mapping to generated/metadata during compilation #18648

- Fixed static test

magento-engcom-team commented:
Hi @pmclain. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.1. 
Please check the release notes for final confirmation.
